libraries/AP_Relay/AP_Relay.cpp

```diff
@@ -128,61 +120,237 @@ AP_Relay::AP_Relay(void)
+        motors->get_legacy_relay_index(rover_relay[0], rover_relay[1], rover_relay[2], rover_relay[3]);
+    }
+#endif
+
+    // Find old default param
+    int8_t default_state = 0; // off was the old behaviour
+    const bool have_default = AP_Param::get_param_by_index(this, 4, AP_PARAM_INT8, &default_state);
+
+    // grab the old values if they were set
+    for (uint8_t i = 0; i < MIN(AP_RELAY_NUM_RELAYS, 6); i++) {
+        if (_params[i].function.configured()) {
+            // Conversion already done, or user has configured manually
+            continue;
+        }
+
+        uint8_t param_index = i;
+        if (i > 3) {
+            // Skip over the old DEFAULT parameter at index 4
+            param_index += 1;
+        }
+
+        int8_t pin = 0;
+        if (!_params[i].pin.configured() && AP_Param::get_param_by_index(this, param_index, AP_PARAM_INT8, &pin) && (pin >= 0)) {
+            // Copy old pin parameter if valid
+            _params[i].pin.set_and_save(pin);
+        }
+
+        // Work out what function this relay should be
+        AP_Relay_Params::FUNCTION new_fun;
+        if (i == chute_relay) {
+            new_fun = AP_Relay_Params::FUNCTION::PARACHUTE;
+
+        } else if (i == ice_relay) {
+            new_fun = AP_Relay_Params::FUNCTION::IGNITION;
+
+        } else if (i == cam_relay) {
+            new_fun = AP_Relay_Params::FUNCTION::CAMERA;
+
+#if APM_BUILD_TYPE(APM_BUILD_Rover)
+        } else if (i == rover_relay[0]) {
+            new_fun = AP_Relay_Params::FUNCTION::BRUSHED_REVERSE_1;
+
+        } else if (i == rover_relay[1]) {
+            new_fun = AP_Relay_Params::FUNCTION::BRUSHED_REVERSE_2;
+
+        } else if (i == rover_relay[2]) {
+            new_fun = AP_Relay_Params::FUNCTION::BRUSHED_REVERSE_3;
+
+        } else if (i == rover_relay[3]) {
+            new_fun = AP_Relay_Params::FUNCTION::BRUSHED_REVERSE_4;
+#endif
+
+        } else {
+            if (_params[i].pin < 0) {
+                // Don't enable as numbered relay if pin is invalid
+                // Other functions should be enabled with a invalid pin
+                // This will result in a pre-arm promoting the user to resolve
+                continue;
+            }
+            new_fun = AP_Relay_Params::FUNCTION::RELAY;
+
+        }
+        _params[i].function.set_and_save(int8_t(new_fun));
+
+
+        // Set the default state
+        if (have_default) {
+            _params[i].default_state.set_and_save(default_state);
+        }
     }
 }
 
-void AP_Relay::set(const uint8_t instance, const bool value)
+void AP_Relay::set_defaults() {
+    const int8_t pins[] = { RELAY1_PIN_DEFAULT,
+                             RELAY2_PIN_DEFAULT,
+                             RELAY3_PIN_DEFAULT,
+                             RELAY4_PIN_DEFAULT,
+                             RELAY5_PIN_DEFAULT,
+                             RELAY6_PIN_DEFAULT };
+
+    for (uint8_t i = 0; i < MIN(uint8_t(AP_RELAY_NUM_RELAYS), ARRAY_SIZE(pins)); i++) {
```

Comment:
Should probably be `ARRAY_SIZE(_params)`

Reply:
fixed